dbms/src/Debug/AstToExecutor.cpp

```diff
@@ -0,0 +1,1516 @@
+            column.name = exprToString(c, {}) + "_" + target_type->getName();
+            column.type = target_type;
+            arguments_types.emplace_back(target_type);
+            argument_columns.emplace_back(column);
+        }
+        auto func_name = getFunctionNameForConstantFolding(expr);
+        if (func_name.empty())
+            return;
+        const auto & function_builder_ptr = FunctionFactory::instance().get(func_name, context);
+        auto function_ptr = function_builder_ptr->build(argument_columns);
+        if (function_ptr->isSuitableForConstantFolding())
+        {
+            Block block_with_constants(argument_columns);
+            ColumnNumbers argument_numbers(arguments_types.size());
+            for (size_t i = 0, size = arguments_types.size(); i < size; i++)
+                argument_numbers[i] = i;
+            size_t result_pos = argument_numbers.size();
+            block_with_constants.insert({nullptr, function_ptr->getReturnType(), "result"});
+            function_ptr->execute(block_with_constants, argument_numbers, result_pos);
+            const auto & result_column = block_with_constants.getByPosition(result_pos).column;
+            if (result_column->isColumnConst())
+            {
+                auto updated_value = (*result_column)[0];
+                tipb::FieldType orig_field_type = expr->field_type();
+                expr->Clear();
+                literalToPB(expr, updated_value, collator_id);
+                expr->clear_field_type();
+                auto * field_type = expr->mutable_field_type();
+                (*field_type) = orig_field_type;
+            }
+        }
+    }
+}
+
+
+void astToPB(const DAGSchema & input, ASTPtr ast, tipb::Expr * expr, uint32_t collator_id, const Context & context)
+{
+    if (ASTIdentifier * id = typeid_cast<ASTIdentifier *>(ast.get()))
+    {
+        auto ft = std::find_if(input.begin(), input.end(), [&](const auto & field) {
+            auto column_name = splitQualifiedName(id->getColumnName());
+            auto field_name = splitQualifiedName(field.first);
+            if (column_name.first.empty())
+                return field_name.second == column_name.second;
+            else
+                return field_name.first == column_name.first && field_name.second == column_name.second;
+        });
+        if (ft == input.end())
+            throw Exception("No such column " + id->getColumnName(), ErrorCodes::NO_SUCH_COLUMN_IN_TABLE);
+        expr->set_tp(tipb::ColumnRef);
+        *(expr->mutable_field_type()) = columnInfoToFieldType((*ft).second);
+        expr->mutable_field_type()->set_collate(collator_id);
+        WriteBufferFromOwnString ss;
+        encodeDAGInt64(ft - input.begin(), ss);
+        expr->set_val(ss.releaseStr());
+    }
+    else if (ASTFunction * func = typeid_cast<ASTFunction *>(ast.get()))
+    {
+        /// aggregation function is handled in Aggregation, so just treated as a column
```

Comment:
how about move this branch into another function?

Reply:
> how about move this branch into another function?

Good idea. I will refactor them all.